app/eth-watcher.hoon

```diff
@@ -0,0 +1,556 @@
+      ==
+    |-  ^+  +>.^$
+    ?~  changes  +>.^$
+    =.  +>.^$
+      (take-event-log (parse-event-log i.changes))
+    $(changes t.changes)
+  ::
+  ::  +take-event-log: obtain changes from event-log
+  ::
+  ++  take-event-log
+    |=  log=event-log
+    ^+  +>
+    ?~  mined.log
+      ~&  %ignoring-unmined-event
+      +>
+    =*  place  u.mined.log
+    ?:  (~(has in heard.eye) block-number.place log-index.place)
+      ?.  removed.u.mined.log
+        ~?  debug=|
+          [%ignoring-duplicate-event tx=transaction-hash.u.mined.log]
+        +>
+      ::  block was reorganized away, so rewind to this block and
+      ::  start syncing again.
+      ::
+      ~&  :*  'removed event!  Perhaps chain has reorganized?'
+              tx-hash=transaction-hash.u.mined.log
+              block-number=block-number.u.mined.log
+              block-hash=block-hash.u.mined.log
+          ==
+      %=    +>
+          rewind-block
+        :-  ~
+        ?~  rewind-block
+          block-number.place
+        (min block-number.place u.rewind-block)
+      ==
+    =.  last-heard-block.eye
+      (max block-number.place last-heard-block.eye)
+    ?:  ?&  (gte block-number.place from-block.eye)
+            ?|  ?=(~ to-block.eye)
+                (lte block-number.place u.to-block.eye)
+            ==
+        ==
+      (put-log log)
+    ~&  :*  %event-block-out-of-range
+            got=block-number.place
+            from=from-block.eye
+            to=to-block.eye
+        ==
+    +>.$
+  ::
+  ::  +restore-block: rewind to block or earlier
+  ::
+  ++  restore-block
+    |=  block=@ud
+    ^+  +>
+    =/  old-qeu  snaps.sap.eye
+    ::  clear history
+    ::
+    =:  snaps.sap.eye       ~
+        count.sap.eye       0
+        latest-block.sap.eye  0
+      ==
+    ::  find a snapshot we can use, remove ones that are too new
+    ::
+    =^  snap=snapshot  +>.$
+      ?:  |(=(~ old-qeu) (lth block last-heard-block:(need ~(top to old-qeu))))
+        [%*(. *snapshot last-heard-block from-block.eye) +>.$]
+      |-  ^-  [snapshot _+>.^$]
+      =^  snap=snapshot  old-qeu
+        ~(get to old-qeu)
+      =:  count.sap.eye       +(count.sap.eye)
+          latest-block.sap.eye  last-heard-block.snap
+          snaps.sap.eye       (~(put to snaps.sap.eye) snap)
+        ==
+      ?:  |(=(~ old-qeu) (lth block last-heard-block:(need ~(top to old-qeu))))
+        [snap +>.^$]
+      $
+    ~&  [%restoring-block block last-heard-block.snap]
+    (restore-snap snap)
+  ::
+  ::  +restore-snap: revert state to snapshot
+  ::
+  ++  restore-snap
+    |=  snap=snapshot
+    ^+  +>
+    ::  notify subscribers
+    ::TODO  be more nuanced about what changed, maybe
+    ::
+    =.  +>.$  (fan-diff snap+snap)
+    ::  restore state and kick new fetch cycle
+    ::
+    %=    get-latest-block
+        last-heard-block.eye  last-heard-block.snap
+        heard.eye         heard.snap
+        logs.eye          logs.snap
+    ==
+  --
+--
```

Comment:
need a trailing newline

Reply:
"Need"? Is there something that would be tripped up by not having trailing newlines?

Reply:
Yeah, isn't that what sunk ~zod one of those times?  We might manage to parse it for gall apps, but at least hoon.hoon really needs a trailing newline or it won't compile.  Plus, a text file with no trailing newline is really malformed.